### DISCUSSION TITLE
How do I programmatically import .osz files?

### DISCUSSION THREAD

**Discussion by nanashi-1** [ppy/osu#19559] [Q&A] [answered]
I'm working to make editing beatmap files externally possible.

Below are the steps on how this feature would work.

1. User will click the button to edit files externally.
2. The files of the beatmap set will be extracted to a folder called "temps".
3. The directory where the beatmap set files are extracted will be presented to the user.
4. Listen when the user edits anything in the directory.
5. When any changes are detected, compress the directory into an osz file then import.

I finished everything but don't know how to import .osz files programmatically.

**Comment by peppy** [comment, marked as answer]
It should not be reimported. The directory should use a file watcher and update resources in the game in real-time (while the editor is open, only). The temporary folder should be removed when exiting the editor.

**Reply by nanashi-1** [reply]
If this is the case, it would be better if it is extracted when the editor is opened rather than being extracted when the button is clicked.

Also, updating the resources in the game in real-time sounds cool. It would be better than osu!stable where every time I make changes, I need to exit the editor and then go back to see the changes. I'll work on it.

**Reply by peppy** [reply]
Because it adds a considerable overhead, it will need to be a toggle.

This is no small feature and it would need to be implemented very well. I'd suggest starting with small pieces (no UI, tests only, showing the process, file watching and realm hook-up working) first.
